### DISCUSSION TITLE
try ... except ... else ... finally 的 finally 真的有需要嗎?

### DISCUSSION THREAD

**Discussion by chia-kai-chang** [1101-GS4719-C/Python-Community-GS4719-C#186] [General]
有同學問了一個很深的問題：
```
try ... except ... else ... finally 的 finally 真的有需要嗎?
因為finally是不論有沒有發生例外都會執行的程式區塊，
那這樣就把finally的程式區塊放到try ... except 整個敘述的
後面就可以了，也可以保證try ... except 執行完畢後，下一步
就會執行到接下來的程式區塊。

那麼還有必要使用finally程式區塊嗎?
```

同學所說的情境如以下的程式：
```
try:
    run_code1()
except TypeError:
    run_code2()
finally:
    other_code()
```

用以下的程式來寫也有相同的效果。
```
try:
    run_code1()
except TypeError:
    run_code2()
other_code()
```

其實如果你的try...except...區塊是放在主程式，確實finally區塊的程式移到try...except...區塊後面也可以有相同的效果。

但是如果你的try...except...區塊是放在function裡面，就會有差別了。

以下的程式都是在function裡面的，所以會看到`return`關鍵字：

看看以下的程式：

程式A：即使發生例外，`other_code()`會在執行後，function才會`return`回主程式
```
try:
    run_code1()
except TypeError:
    run_code2()
    return None   # The finally block is run before the method returns
finally:
    other_code()
```

程式B：如果發生例外，`other_code()`在try...except...區塊外面，所以直到function `return`，`other_code()`都沒有機會執行
```
try:
    run_code1()
except TypeError:
    run_code2()
    return None   

other_code()  # This doesn't get run if there's an exception.
```

其他可能會導致差異的情況：

1. 在except區塊內拋出例外錯誤
2. 在執行`run_code1()`時發生非`TypeError`的例外錯誤
3. 其他流程控制(如：`continue`或`break`)所導致的情況

😂

**Comment by xCk27x** [comment]
所以意思是 finally 後的敘述算是一個強制執行的區塊 就算 try except 途中出現 return 也依然會執行

**Reply by enoch-lightning** [reply]
@xCk27x 你這樣講好像蠻合理的，運作方式應該就是如你講的這樣吧！

**Reply by chia-kai-chang** [reply]
> 所以意思是 finally 後的敘述算是一個強制執行的區塊 就算 try except 途中出現 return 也依然會執行

@xCk27x 是的，可以確保`finally`裡面的敘述可以被執行。

**Reply by enoch-lightning** [reply]
> 所以意思是 finally 後的敘述算是一個強制執行的區塊 就算 try except 途中出現 return 也依然會執行
@xCk27x 是的，可以確保finally裡面的敘述可以被執行。

原來例外處理，也就是異常處理的try...except...finally的finally是try...except...finally中一定會被執行到的最後區域。這樣就不會因為進入例外情況except，接而break(跳出)或return(回傳給呼叫者(caller))，導致一定要執行的程式不被執行。
